### PR TITLE
Google Classroom Roster Integration

### PR DESCRIPTION
**Pull request by srinjoym**
Roster import with syncing from Google Classroom! ✨ 

Depends on #1827 

Todo:
- [x] Full Test Coverage
- [x] Create feature flag
- [ ] Add Experimental Feedback flag in #1835 
- [ ] Merge #1827

### CODE REVIEW

File: app/views/orgs/rosters/select_google_classroom.html.erb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    <div class="my-3">
+      <h2 class="h3 mb-2">Select the Google Classroom roster you want to import</h2>
+
+      <div class="Box overflow-auto">
+        <ul>
+          <% @google_classroom_courses.each do |course| %>
+            <li class="Box-row">
+              <label>
+                <input type="checkbox" class="mr-2" checked="checked" aria-describedby="help-text-for-checkbox">
```
Comment (brandonrosage) [inline app/views/orgs/rosters/select_google_classroom.html.erb]: Since you can only import (and associate with) one Google Classroom (in the interest of syncing later), I suspect these should be radio buttons.

File: app/views/orgs/rosters/select_google_classroom.html.erb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+                <%= course.name %>
+                <%# <%= link_to course.name, {controller: "rosters", action: :import_from_google_classroom, course_id: course.id}, method: :patch %>
+              </label>
+            </li>
+          <% end %>
+        </ul>
+      </div>
+    </div>
+
+    <div class="f4 text-gray text-italic my-3">or <a href="" class="btn-link">manually add students</a></div>
```
Comment (brandonrosage) [inline app/views/orgs/rosters/select_google_classroom.html.erb]: This needs to be linked back to the default "manual" view for creating a roster.

File: .env-example [hunk 3]
```diff
@@ -1,6 +1,10 @@
 GITHUB_CLIENT_ID=
 GITHUB_CLIENT_SECRET=
 
+# Used for Google Classroom Roster Integration
+GOOGLE_CLIENT_ID=
```
Comment (srinjoym) [inline .env-example]: Used for communicating with Google Classroom API

File: app/models/roster/creator.rb [hunk 4]
```diff
@@ -57,12 +57,15 @@ def initialize(organization:, identifier_name:, **options)
...
     #
+    # rubocop:disable Metrics/MethodLength
     def perform
       ensure_organization_does_not_have_roster!
 
       ActiveRecord::Base.transaction do
         @roster = Roster.new(identifier_name: @identifier_name)
-        add_identifiers_to_roster(@options[:identifiers]) if @options.key?(:identifiers)
+
+        google_ids = @options[:google_user_ids].nil? ? [] : @options[:google_user_ids]
```
Comment (srinjoym) [inline app/models/roster/creator.rb]: Optional `:google_user_ids` key is used to associate Google Classroom student ID with the Roster Entry

File: app/models/roster/creator.rb [hunk 5]
```diff
@@ -72,13 +75,15 @@ def perform
...
 
     private
 
-    def add_identifiers_to_roster(raw_identifiers_string)
+    def add_identifiers_to_roster(raw_identifiers_string, google_ids: [])
       identifiers = raw_identifiers_string.split("\r\n").reject(&:blank?).uniq
-      identifiers.each do |identifier|
-        @roster.roster_entries << RosterEntry.new(identifier: identifier)
+
+      identifiers.zip(google_ids).each do |identifier, google_user_id|
```
Comment (srinjoym) [inline app/models/roster/creator.rb]: If there are no `google_ids` this just puts a nil for each roster entry

File: config/initializers/google_classroom_client.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    scope = [
+      Google::Apis::ClassroomV1::AUTH_CLASSROOM_COURSES_READONLY,
+      Google::Apis::ClassroomV1::AUTH_CLASSROOM_ROSTERS_READONLY
+    ]
+
+    client_id = Rails.application.secrets.google_client_id
+    client_secret = Rails.application.secrets.google_client_secret
+    client = Google::Auth::ClientId.new(client_id, client_secret)
+
+    token_store = Google::Auth::Stores::RedisTokenStore.new(redis: GitHubClassroom.redis)
```
Comment (srinjoym) [inline config/initializers/google_classroom_client.rb]: Using our existing redis store to store Google Classroom user tokens

**Comment by srinjoym** [comment]
@BenEmdon When you get a chance could you do a preliminary review of this PR? Sorry for the huge size, the actual changes are all pretty much in the RostersController but the cassettes and gems increased the number of changed files. Working on a follow up PR with UI updates for this + throughout Classroom in #1835

File: app/controllers/orgs/rosters_controller.rb [hunk 7]
```diff
@@ -146,6 +175,60 @@ def download_roster
     # rubocop:enable Metrics/MethodLength
     # rubocop:enable Metrics/AbcSize
 
+    def select_google_classroom
```
Comment (srinjoym) [inline app/controllers/orgs/rosters_controller.rb]: Route used when displaying list of Google Classrooms to select

File: Gemfile [hunk 8]
```diff
@@ -44,7 +46,6 @@ gem "local_time", "~> 2.0"
 gem "octicons_helper", "~> 2.1"
 gem "octokit", github: "octokit/octokit.rb", ref: "ffae5ddd6fd9da6f4538ca7adeb5d1768011610b"
 gem "octopoller",      "~> 0.1"
-gem "omniauth",        "~> 1.6", ">= 1.6.1"
```
Comment (BenEmdon) [inline Gemfile]: Don't we need omniauth?

File: app/views/orgs/rosters/_google_classroom_collection.html.erb [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+                new_roster_path,
+                class: "btn-link"
+    %>
+  </div>
+
+  <div class="d-flex flex-items-center border-top pt-5">
+    <%= submit_tag 'Import roster', class: 'btn btn-primary mr-3' %>
+    <%= link_to 'Skip', organization_path(current_organization), class: 'btn-link' %>
+  </div>
+<% end %>
```
Comment (BenEmdon) [inline app/views/orgs/rosters/_google_classroom_collection.html.erb]: Nit: breakline before the EOF
```suggestion
<% end %>

```

File: package-lock.json [hunk 10]
```diff
@@ -0,0 +1,945 @@
+{
```
Comment (BenEmdon) [inline package-lock.json]: How come we haven't had a `package-lock.json` until now?

File: config/initializers/google_classroom_client.rb (hunk 6 above)
Comment (BenEmdon) [reply config/initializers/google_classroom_client.rb]: Why did you decide to use Redis instead of Postgres?

File: app/controllers/orgs/rosters_controller.rb [hunk 11]
```diff
@@ -146,6 +175,60 @@ def download_roster
...
+    # rubocop:disable Metrics/AbcSize
+    def search_google_classroom
+      courses_found = fetch_all_google_classrooms.select do |course|
+        course.name.downcase.include? params[:query].downcase
+      end
+
+      respond_to do |format|
+        format.html do
+          render partial: "orgs/rosters/google_classroom_collection",
+                 locals: { courses: courses_found }
```
[the remark this answers is not in the document]
Comment (srinjoym) [reply app/controllers/orgs/rosters_controller.rb]: Tried to add this but it hits a Linter Error for AlignHash https://github.com/rubocop-hq/rubocop/blob/master/manual/cops_layout.md#layoutalignhash.

I think it makes sense to update the rubocop rule, didn't want to turn it off for everything because we still want the check but I'm sure there's some configuration we can find in a future PR.

File: app/controllers/orgs/rosters_controller.rb [hunk 12]
```diff
@@ -228,6 +311,88 @@ def get_user_to_group_hash(grouping)
...
+      @google_classroom_service.client_options.application_name = "GitHub Classroom"
+      @google_classroom_service.authorization = user_google_classroom_credentials
+    end
+
+    # Helper method for getting current user's google classroom credentials
+    def user_google_classroom_credentials
+      google_classroom_client = GitHubClassroom.google_classroom_client
+      user_id = current_user.github_user.login
+
+      google_classroom_client.get_credentials(user_id, request)
```
[the remark this answers is not in the document]
Comment (srinjoym) [reply app/controllers/orgs/rosters_controller.rb]: This is the key that is stored in Redis for looking up the Google Classroom user's credentials for signing into the API

File: app/controllers/orgs/rosters_controller.rb [hunk 13]
```diff
@@ -146,6 +175,60 @@ def download_roster
...
+        redirect_to roster_path(current_organization)
+      else
+        current_organization.update_attributes!(google_course_id: params[:course_id])
+        add_google_classroom_students(students)
+      end
+    end
+
+    # rubocop:disable Metrics/AbcSize
+    def sync_google_classroom
+      return if current_organization.google_course_id.nil?
```
[the remark this answers is not in the document]
Comment (srinjoym) [reply app/controllers/orgs/rosters_controller.rb]: That's a good question, not sure what to do here because if we're trying to sync a google classroom roster and we don't have an ID it's in a pretty broken state 🤔 I think if we just flash an error it should be fine?

File: app/controllers/orgs/rosters_controller.rb [hunk 14]
```diff
@@ -146,6 +175,60 @@ def download_roster
...
+
+    def import_from_google_classroom
+      students = list_google_classroom_students(params[:course_id])
+      return if students.nil?
+
+      if students.blank?
+        flash[:warning] = "No students were found in your Google Classroom. Please add students and try again."
+        redirect_to roster_path(current_organization)
+      else
+        current_organization.update_attributes!(google_course_id: params[:course_id])
```
[the remark this answers is not in the document]
Comment (srinjoym) [reply app/controllers/orgs/rosters_controller.rb]: Great catch, just updated this to not skip vallidations

File: app/controllers/orgs/rosters_controller.rb [hunk 15]
```diff
@@ -36,8 +59,9 @@ def new
     def create
       result = Roster::Creator.perform(
         organization: current_organization,
-        identifier_name: params[:identifier_name],
-        identifiers: params[:identifiers]
+        identifier_name: "Identifiers",
```
[the remark this answers is not in the document]
Comment (srinjoym) [reply app/controllers/orgs/rosters_controller.rb]: Great idea, just moved it downstream to `Roster::Creator`

File: package-lock.json (hunk 10 above)
Comment (srinjoym) [reply package-lock.json]: Whoops, must've used npm instead of yarn at some point, deleted the file

**Comment by srinjoym** [comment]
@BenEmdon Just addressed a bunch of the comments, thanks for the great feedback! Couldn't reply to the PG vs Redis comment so I'll just post here. The Google APIs authentication module `Google::Auth` supports two main types of token storage: Redis and Files. I figured since we already have a Redis store we're using we could just use that for the user credentials as well. Here's some more documentation on `Google::Auth`: https://www.rubydoc.info/github/google/google-auth-library-ruby/Google/Auth/TokenStore

File: app/controllers/orgs/rosters_controller.rb [hunk 16]
```diff
@@ -76,7 +99,7 @@ def link
       user_id = params[:user_id].to_i
       raise ActiveRecord::ActiveRecordError unless unlinked_user_ids.include?(user_id)
 
-      current_roster_entry.update_attributes!(user_id: user_id)
+      current_roster_entry.update_attributes(user_id: user_id)
```
Comment (BenEmdon) [inline app/controllers/orgs/rosters_controller.rb]: `update_attribute` skips validations. Use `update` instead to _not_ skip validations. 
See https://davidverhasselt.com/set-attributes-in-activerecord/ for details.
```suggestion
      current_roster_entry.update(user_id: user_id)
```

File: app/controllers/orgs/rosters_controller.rb [hunk 17]
```diff
@@ -146,6 +174,58 @@ def download_roster
...
+
+    def import_from_google_classroom
+      students = list_google_classroom_students(params[:course_id])
+      return if students.nil?
+
+      if students.blank?
+        flash[:warning] = "No students were found in your Google Classroom. Please add students and try again."
+        redirect_to roster_path(current_organization)
+      else
+        current_organization.update_attributes(google_course_id: params[:course_id])
```
Comment (BenEmdon) [inline app/controllers/orgs/rosters_controller.rb]: `update_attribute` skips validations. Use `update` instead to _not_ skip validations. 
See https://davidverhasselt.com/set-attributes-in-activerecord/ for details.
```suggestion
        current_organization.update(google_course_id: params[:course_id])
```

File: app/controllers/orgs/rosters_controller.rb [hunk 18]
```diff
@@ -146,6 +174,58 @@ def download_roster
...
+
+      new_student_ids = latest_student_ids - current_student_ids
+      new_students = latest_students.select { |s| new_student_ids.include? s.user_id }
+
+      add_google_classroom_students(new_students)
+    end
+    # rubocop:enable Metrics/AbcSize
+
+    def unlink_google_classroom
+      current_organization.update_attributes!(google_course_id: nil)
```
Comment (BenEmdon) [inline app/controllers/orgs/rosters_controller.rb]: `update_attribute` skips validations. Use `update` instead to _not_ skip validations. 
See https://davidverhasselt.com/set-attributes-in-activerecord/ for details.
```suggestion
      current_organization.update!(google_course_id: nil)
```

File: app/controllers/orgs/rosters_controller.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (BenEmdon) [reply app/controllers/orgs/rosters_controller.rb]: Flashing an error sounds like a good idea 👍

File: app/controllers/orgs/rosters_controller.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (BenEmdon) [reply app/controllers/orgs/rosters_controller.rb]: If a user changed their name would this break?

[the remark this answers is not in the document]
Comment (thebradbain) [reply app/controllers/orgs/rosters_controller.rb]: @srinjoym is there any reason we can't simply store a classroom user's `id` or, if it needs to use the github_user object, the (unchanging?) `user,github_user.user_id` as the key?

File: app/controllers/orgs/rosters_controller.rb [hunk 19]
```diff
@@ -228,6 +308,88 @@ def get_user_to_group_hash(grouping)
...
     end
+
+    # Returns name of the linked google course to current organization (for syncing rosters)
+    def current_organization_google_course_name
+      return nil if current_organization.google_course_id.nil?
+      authorize_google_classroom
+      course = @google_classroom_service.get_course(current_organization.google_course_id)
+      return course.name unless course.nil?
+      nil
+    rescue Google::Apis::Error
```
Comment (stephaniegiang) [inline app/controllers/orgs/rosters_controller.rb]: Needs another indentation

**Comment by stephaniegiang** [comment]
@srinjoym won't be working on this PR anymore. So I'm going to take over!

File: app/models/roster/creator.rb (hunk 5 above)
Comment (stephaniegiang) [reply app/models/roster/creator.rb]: I believe so. If there is an existing roster and we import Google Classroom, there will be more roster entries than the google_ids.

**Comment by d12** [comment]
Ping me when this is ready for review!

File: app/models/roster/creator.rb (hunk 5 above)
Comment (stephaniegiang) [reply app/models/roster/creator.rb]: Just talked to @srinjoym and there would not be a chance that these are uneven. We are grabbing the google classroom identifiers and ids to the roster entries.

**Comment by stephaniegiang** [comment]
@d12 ready for review! Thanks!

File: app/controllers/orgs/rosters_controller.rb [hunk 20]
```diff
@@ -146,6 +174,61 @@ def download_roster
     # rubocop:enable Metrics/MethodLength
     # rubocop:enable Metrics/AbcSize
 
+    def select_google_classroom
+      @google_classroom_courses = fetch_all_google_classrooms
+    end
+
+    def search_google_classroom
+      courses_found = fetch_all_google_classrooms.select do |course|
+        course.name.downcase.include? params[:query].downcase
```
Comment (d12) [inline app/controllers/orgs/rosters_controller.rb]: Can we use a SQL `LIKE` condition here instead of doing filtering in Ruby?

File: app/controllers/orgs/rosters_controller.rb [hunk 21]
```diff
@@ -146,6 +174,61 @@ def download_roster
...
+        format.html do
+          render partial: "orgs/rosters/google_classroom_collection",
+                 locals: { courses: courses_found }
+        end
+      end
+    end
+
+    def import_from_google_classroom
+      students = list_google_classroom_students(params[:course_id])
+      return if students.nil?
```
Comment (d12) [inline app/controllers/orgs/rosters_controller.rb]:
```suggestion
      return unless students
```

File: app/controllers/orgs/rosters_controller.rb [hunk 22]
```diff
@@ -228,6 +311,88 @@ def get_user_to_group_hash(grouping)
...
+      nil
+    rescue Google::Apis::Error
+      nil
+    end
+
+    # Returns list of students in a google classroom with error checking
+    # rubocop:disable Metrics/MethodLength
+    def list_google_classroom_students(course_id)
+      response = @google_classroom_service.list_course_students(course_id)
+      response.students ||= [] # Set to empty array if no students in course
```
Comment (d12) [inline app/controllers/orgs/rosters_controller.rb]: Do you mind adding a slightly more explicit `response.students` line under this to show that's the return value? it's a little tricky following how `response.students ||= []` returns response.students if it exists.

File: app/controllers/orgs/rosters_controller.rb [hunk 23]
```diff
@@ -146,6 +174,61 @@ def download_roster
...
+        redirect_to roster_path(current_organization)
+      else
+        current_organization.update(google_course_id: params[:course_id])
+        add_google_classroom_students(students)
+      end
+    end
+
+    # rubocop:disable Metrics/AbcSize
+    def sync_google_classroom
+      if current_organization.google_course_id.nil?
```
Comment (d12) [inline app/controllers/orgs/rosters_controller.rb]:
```suggestion
     unless current_organization.google_course_id
```

File: app/controllers/orgs/rosters_controller.rb [hunk 24]
```diff
@@ -146,6 +174,61 @@ def download_roster
...
+        flash[:error] = "No Google Classroom has been linked. Please link Google Classroom."
+        return;
```
Comment (d12) [inline app/controllers/orgs/rosters_controller.rb]: :fire: ; :fire:

File: app/controllers/orgs/rosters_controller.rb [hunk 25]
```diff
@@ -228,6 +311,88 @@ def get_user_to_group_hash(grouping)
...
+
+    # Add Google Classroom students to roster
+    def add_google_classroom_students(students)
+      names = students.map { |s| s.profile.name.full_name }
+      user_ids = students.map(&:user_id)
+      params[:identifiers] = names.join("\r\n")
+      params[:google_user_ids] = user_ids
+
+      if current_roster.nil?
+        create
```
Comment (d12) [inline app/controllers/orgs/rosters_controller.rb]: This pattern of calling controller actions from within controller actions can lead to some weird side effects. Can we abstract the logic you need into a method and then call that method from both controller actions?

File: app/controllers/orgs/rosters_controller.rb [hunk 26]
```diff
@@ -228,6 +311,88 @@ def get_user_to_group_hash(grouping)
...
+      else
+        add_students
```
Comment (d12) [inline app/controllers/orgs/rosters_controller.rb]: Same comment here

File: app/controllers/orgs/rosters_controller.rb [hunk 27]
```diff
@@ -228,6 +311,88 @@ def get_user_to_group_hash(grouping)
...
+      end
+
+      courses
+    end
+
+    # Authorizes current user through Google and sets google_classroom_service
+    # Used as a before_action before routes which require Google authorization
+    def authorize_google_classroom
+      google_classroom_client = GitHubClassroom.google_classroom_client
+      if user_google_classroom_credentials.nil?
```
Comment (d12) [inline app/controllers/orgs/rosters_controller.rb]:
```suggestion
      unless user_google_classroom_credentials
```

File: db/schema.rb [hunk 28]
```diff
@@ -110,6 +110,7 @@
     t.integer "max_members"
     t.boolean "students_are_repo_admins", default: false, null: false
     t.boolean "invitations_enabled", default: true
+    t.integer "max_teams"
```
Comment (d12) [inline db/schema.rb]: I thought this already merged? You might have some other commits in this PR.

File: config/initializers/rack_rewrite.rb [hunk 29]
```diff
@@ -2,7 +2,7 @@
 
 # URL for Nuts Server deployed on Heroku, redirects download/update requests to latest release
 # on GitHub Releases with platform parameters
-RELEASE_SERVER_URL = "http://classroom-assistant-releases.herokuapp.com"
```
Comment (d12) [inline config/initializers/rack_rewrite.rb]: Can this move into another PR incase we need to revert this?

File: db/schema.rb (hunk 28 above)
Comment (stephaniegiang) [reply db/schema.rb]: Oops I will fix that.

**Comment by stephaniegiang** [comment]
@d12 ready for another round of review

File: .env-example (hunk 3 above)
Comment (d12) [inline .env-example]: Can we add something to the README about this in the section about filling out the .env? Mainly is this optional? I have no idea how I'd go about getting a client ID and secret.

File: app/controllers/orgs/rosters_controller.rb [hunk 30]
```diff
@@ -146,6 +174,61 @@ def download_roster
     # rubocop:enable Metrics/MethodLength
     # rubocop:enable Metrics/AbcSize
 
+    def select_google_classroom
+      @google_classroom_courses = fetch_all_google_classrooms
+    end
+
+    def search_google_classroom
+      courses_found = fetch_all_google_classrooms.select do |course|
+        course.where("name ILIKE ?", "#{params[:query]}#")
```
Comment (d12) [inline app/controllers/orgs/rosters_controller.rb]: Ah sorry, I think I may have left bad feedback here, I thought this returned an AR relation. It looks like this fires a SQL query per course which isn't great for performance. If this is an array and the data is already loaded, lets just do the filtering in memory here. Something like `course.name.downcase.include?(params[:query].downcase)`

File: app/controllers/orgs/rosters_controller.rb [hunk 31]
```diff
@@ -228,6 +311,89 @@ def get_user_to_group_hash(grouping)
 
       mapping
     end
+
+    # Returns name of the linked google course to current organization (for syncing rosters)
+    def current_organization_google_course_name
+      return nil if current_organization.google_course_id.nil?
```
Comment (d12) [inline app/controllers/orgs/rosters_controller.rb]: If you find yourself writing `.nil?` in a conditional, there's usually a cleaner way to write it 😅 

```suggestion
     return unless current_organization.google_course_id
```

File: app/controllers/orgs/rosters_controller.rb [hunk 32]
```diff
@@ -228,6 +311,89 @@ def get_user_to_group_hash(grouping)
...
+      authorize_google_classroom
+      course = @google_classroom_service.get_course(current_organization.google_course_id)
+      return course.name unless course.nil?
+      nil
```
Comment (d12) [inline app/controllers/orgs/rosters_controller.rb]: This is a little odd, the fall through case of the method is the error case, and we bail early with the expected return type earlier in the method (320). People usually skim to the bottom of the method to see what it returns, so this is a little confusing.

Instead, we could replace 320 and 321 with `return unless course`, and `course.name` below.

 If we want to take it one step further, we can use the [safe navigation operator](https://stackoverflow.com/questions/36812647/what-does-ampersand-dot-mean-in-ruby) here and replace 320 and 321 with 

```ruby
course&.name
```

As an example:

```ruby
irb(main):004:0> (-5).abs
=> 5

irb(main):005:0> nil.abs
NoMethodError: undefined method `abs' for nil:NilClass
	from (irb):5
	from /usr/bin/irb:11:in `<main>'

irb(main):006:0> nil&.abs
=> nil

irb(main):007:0> (-5)&.abs
=> 5
```

File: app/models/roster/creator.rb [hunk 33]
```diff
@@ -45,24 +47,26 @@ def failed?
...
+    # rubocop:disable Metrics/MethodLength
     def perform
       ensure_organization_does_not_have_roster!
 
       ActiveRecord::Base.transaction do
-        @roster = Roster.new(identifier_name: @identifier_name)
-        add_identifiers_to_roster(@options[:identifiers]) if @options.key?(:identifiers)
+        @roster = Roster.new(identifier_name: DEFAULT_IDENTIFIER_NAME)
+
+        google_ids = @options[:google_user_ids].nil? ? [] : @options[:google_user_ids]
```
Comment (d12) [inline app/models/roster/creator.rb]:
```suggestion
        google_ids = @options[:google_user_ids] || []
```

File: config/routes.rb [hunk 34]
```diff
@@ -80,6 +84,11 @@
           patch :delete_entry
           patch :add_students
           patch :remove_organization
+          patch :import_from_google_classroom
+          patch :sync_google_classroom
+          patch :unlink_google_classroom
+          get   :select_google_classroom, as: "select_google_classroom"
```
Comment (d12) [inline config/routes.rb]: Do we need this `as:`?

File: .env-example (hunk 3 above)
Comment (stephaniegiang) [reply .env-example]: Is it possible to add this later? This google integration PR is not quite ready as I need to polish it a bit more (duplicate roster entries and allow teachers to import different identifiers).

I'll also make a separate PR for this. This PR is quite large and I'm getting lost in the files changed already 😂

File: config/routes.rb (hunk 34 above)
Comment (stephaniegiang) [reply config/routes.rb]: No we don't. I'll remove it.

**Comment by stephaniegiang** [comment]
@d12 Ready for another review!

File: app/controllers/orgs/rosters_controller.rb [hunk 35]
```diff
@@ -109,9 +132,14 @@ def delete_entry
     # rubocop:disable Metrics/AbcSize
     def add_students
       identifiers = params[:identifiers].split("\r\n").reject(&:blank?).uniq
+      google_ids = params[:google_user_ids] ? params[:google_user_ids] : []
```
Comment (d12) [inline app/controllers/orgs/rosters_controller.rb]:
```suggestion
      google_ids = params[:google_user_ids] || []
```

File: package.json [hunk 36]
```diff
@@ -9,7 +9,7 @@
     "jquery-readyselector": "git+https://git@github.com/Verba/jquery-readyselector.git",
     "jquery-ujs": " ~1.2.2",
     "jquery.turbolinks": "git+https://git@github.com/kossnocorp/jquery.turbolinks#2.1.0",
-    "primer": "^10.10.4",
+    "primer": "^10.10.5",
```
Comment (d12) [inline package.json]: Why was this change necessary?

File: db/schema.rb (hunk 28 above)
Comment (d12) [reply db/schema.rb]: It looks like this reappeared again

File: spec/controllers/orgs/rosters_controller_spec.rb [hunk 37]
```diff
@@ -135,8 +117,7 @@
           before do
             post :create, params: {
               id: organization.slug,
-              identifiers: "    \r\n ",
-              identifier_name: "emails"
```
Comment (d12) [inline spec/controllers/orgs/rosters_controller_spec.rb]: There are a bunch of changes here to identifier_name stuff, didn't this happen in a previous PR? If not, can we move to a different PR? It's unrelated to the Google Classroom stuff and this PR is already massive.

File: spec/controllers/orgs/rosters_controller_spec.rb [hunk 38]
```diff
@@ -261,6 +242,189 @@
...
+      end
+    end
+
+    context "with flipper disabled" do
+      before do
+        get :show, params: { id: organization.slug }
+      end
+
+      it "404s" do
+        expect(response).to have_http_status(:not_found)
```
Comment (d12) [inline spec/controllers/orgs/rosters_controller_spec.rb]: Why is `:show` 404ing here? Shouldn't `show` render fine regardless of whether or not the flipper is enabled?

File: spec/controllers/orgs/rosters_controller_spec.rb [hunk 39]
```diff
@@ -684,4 +848,415 @@
...
+                id: organization.slug
+              }
+            end
+
+            it "doesn't add any students" do
+              expect(organization.roster.roster_entries.count).to eq(1)
+            end
+          end
+
+          context "classroom has a linked google course" do
```
Comment (d12) [inline spec/controllers/orgs/rosters_controller_spec.rb]: Amazing tests in this context :sparkles:

File: spec/controllers/orgs/rosters_controller_spec.rb (hunk 38 above)
Comment (stephaniegiang) [reply spec/controllers/orgs/rosters_controller_spec.rb]: Yeah we should be linking to the Google Classroom setup instead. Not sure why this is passing. Just tested it manually and it isn't 404ing 😕

Comment (d12) [reply spec/controllers/orgs/rosters_controller_spec.rb]: Could you dig into why that's happening in the test? That's suspicious...

Comment (stephaniegiang) [reply spec/controllers/orgs/rosters_controller_spec.rb]: This test was passing because we did not enable the `:student_identifier` flag. We should really get rid of all the feature flags that been at 100% for a while.

Comment (stephaniegiang) [reply spec/controllers/orgs/rosters_controller_spec.rb]: We also already have a test covering the google class setup route. So I'll just modify this test to pass if the `:google_classroom_roster_import` is disabled.

File: package.json (hunk 36 above)
Comment (stephaniegiang) [reply package.json]: Ooops I'll remove that

**Comment by stephaniegiang** [comment]
@d12 ready for another round of review!

File: db/schema.rb [hunk 40]
```diff
@@ -163,6 +163,17 @@
     t.index ["user_id"], name: "index_invite_statuses_on_user_id"
   end
 
+  create_table "lti_configurations", force: :cascade do |t|
```
Comment (stephaniegiang) [inline db/schema.rb]: Not sure why everytime I pull master, this file always comes back. But I'll fix this 😄

**Comment by d12** [comment]
Teachers are very excited for this ship! Let me know how I can help this move forward.

**Comment by stephaniegiang** [comment]
Just pulled in master, so hopefully if the tests pass, another round of reviews. Then we can merge!

**Comment by stephaniegiang** [comment]
@d12 officially ready for another review!

File: config/initializers/rack_rewrite.rb [hunk 41]
```diff
@@ -2,7 +2,7 @@
 
 # URL for Nuts Server deployed on Heroku, redirects download/update requests to latest release
 # on GitHub Releases with platform parameters
-RELEASE_SERVER_URL = "https://classroom-assistant-releases.herokuapp.com"
+RELEASE_SERVER_URL = "http://classroom-assistant-releases.herokuapp.com"
```
Comment (d12) [inline config/initializers/rack_rewrite.rb]: Why are we making this change?

Comment (stephaniegiang) [reply config/initializers/rack_rewrite.rb]: Oopps the merge of master brought that back

**Comment by stephaniegiang** [comment]
PR will be deployed once we get proper google credientials
